File: 2025/04/main.py
def search(grid,max_adjacent,part2=False):
    adjacency = [(-1,-1),(-1,0),(-1,1),(0,-1),(0,1),(1,-1),(1,0),(1,1)]
    found,found_coords = 0,set()
    while True:
        for x,y in grid:
            adjacent_count = sum(1 for dx,dy in adjacency if (x+dx,y+dy) in grid)
            if adjacent_count < max_adjacent:
                found_coords.add((x,y))
        if not part2:
            return len(found_coords)
        if len(found_coords) == 0:
            return found
        found += len(found_coords)
        grid -= found_coords
        found_coords = set()

if __name__ == "__main__":
    grid = set()
    for y,line in enumerate(open("input.txt")):
        for x,char in enumerate(line.strip()):
            if char == "@":
                grid.add((x,y))
    print("Part 1:",search(grid.copy(),4))
    print("Part 2:",search(grid.copy(),4,True))

File: 2025/04/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Coord struct {
	x, y int
}

var adjacency = []Coord{
	{-1, -1}, {-1, 0}, {-1, 1},
	{0, -1}, {0, 1},
	{1, -1}, {1, 0}, {1, 1},
}

func copyGrid(grid map[Coord]struct{}) map[Coord]struct{} {
	newGrid := make(map[Coord]struct{}, len(grid))
	for k := range grid {
		newGrid[k] = struct{}{}
	}
	return newGrid
}

func countAdjacent(coord Coord, grid map[Coord]struct{}) int {
	count := 0
	for _, adj := range adjacency {
		if _, exists := grid[Coord{coord.x + adj.x, coord.y + adj.y}]; exists {
			count++
		}
	}
	return count
}

func search(grid map[Coord]struct{}, maxAdjacent int, part2 bool) int {
	found := 0

	for {
		foundCoords := make(map[Coord]struct{})

		for coord := range grid {
			if countAdjacent(coord, grid) < maxAdjacent {
				foundCoords[coord] = struct{}{}
			}
		}

		if !part2 {
			return len(foundCoords)
		}

		if len(foundCoords) == 0 {
			return found
		}

		found += len(foundCoords)
		for coord := range foundCoords {
			delete(grid, coord)
		}
	}
}

func main() {
	data, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	grid := make(map[Coord]struct{})
	for y, line := range lines {
		for x, char := range line {
			if char == '@' {
				grid[Coord{x, y}] = struct{}{}
			}
		}
	}

	fmt.Println("Part 1:", search(copyGrid(grid), 4, false))
	fmt.Println("Part 2:", search(copyGrid(grid), 4, true))
}
